vochain/processarchive/processarchive.go

```diff
@@ -0,0 +1,140 @@
+package processarchive
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"go.vocdoni.io/dvote/data"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/types"
+	"go.vocdoni.io/dvote/vochain"
+	"go.vocdoni.io/proto/build/go/models"
+)
+
+type ProcessArchive struct {
+	vochain     *vochain.BaseApplication
+	ipfs        *data.IPFSHandle
+	storage     *jsonStorage
+	pprocs      []*Process
+	publish     chan (bool)
+	lastUpdate  time.Time
+	close       chan (bool)
+	publishLock sync.Mutex
+}
+
+type Process struct {
+	Process *models.Process
+	Votes   uint32
+}
+
+type jsonStorage struct {
+	datadir string
+	lock    sync.RWMutex
+}
+
+func NewJsonStorage(datadir string) (*jsonStorage, error) {
+	err := os.MkdirAll(datadir, 0o750)
+	if err != nil {
+		return nil, err
+	}
+	return &jsonStorage{datadir: datadir}, nil
+}
+
+func (js *jsonStorage) AddProcess(p *Process) error {
+	if p == nil || p.Process == nil || len(p.Process.ProcessId) != types.ProcessIDsize {
+		return fmt.Errorf("process not valid")
+	}
+	data, err := json.Marshal(p)
+	if err != nil {
+		return err
+	}
+	js.lock.Lock()
+	defer js.lock.Unlock()
+	return os.WriteFile(filepath.Join(js.datadir, fmt.Sprintf("%x", p.Process.ProcessId)), data, 0o644)
+}
+
+// NewProcessArchive creates a new instance of the process archiver.
+// It will subscribe to Vochain events and perform the process archival.
+// JSON files (one per process) will be stored within datadir.
+// The key parameter must be either a valid IPFS base64 encoded private key
+// or empty (a new key will be generated).
+// If ipfs is nil, only JSON archive storage will be performed.
+func NewProcessArchive(v *vochain.BaseApplication, ipfs *data.IPFSHandle,
+	datadir, key string) (*ProcessArchive, error) {
+	js, err := NewJsonStorage(datadir)
+	if err != nil {
+		return nil, fmt.Errorf("could not create process archive: %w", err)
+	}
+	ir := &ProcessArchive{
+		vochain: v,
+		ipfs:    ipfs,
+		storage: js,
+		publish: make(chan (bool), 10),
+		close:   make(chan (bool), 1),
```

Comment:
I think it would be better to make NewProcesArchive take a context, and then plumb it down to Publish. Practically the same in practice, but more idiomatic, and we don't need to remember to call the Close method and write the channel stuff ourselves.